Changelog 162 (#142)
intellij-rust/intellij-rust.github.io: _posts/2021-12-20-changelog-162.markdown

---
layout: post
title: "IntelliJ Rust Changelog #162"
date: 2021-12-20 13:00:00 +0300
---

## New Features

* [#8145] Take into account [const generics defaults] in const evaluation

> * [#8070] Suggest private items when completing code in the debugger

Reviewer: ```suggestion
* [#8070] Suggest private items when completing code inside `New Watch` and `Evaluate` in [Debug](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-debug-tool-window.html) tool window
```

> * [#8070] Suggest private items when completing code in the debugger

Reviewer: Don't forget to add a gif as we discussed in person

* [#8192] Parse and annotate [`~const` syntax]

## Fixes

* Procedural macro expansion fixes:

>   * [#8271] Fix expansion of `sqlx` procedural macros with `offline` feature

Reviewer: Maybe add a link to library?

>   * [#8271] Fix expansion of `sqlx` procedural macros with `offline` feature

Reviewer: Do I understand correctly that this fix is not only about `sqlx` library but about any proc macro that writes code using `CARGO_TARGET_DIR` variable?
If it's true, let's write it and make "`sqlx` procedural macros with `offline` feature" as example

>   * [#8233] fix expansion of proc macros that weirdly mix token spans (e.g. `yew::html!`)

Reviewer: ```suggestion
  * [#8233] Fix expansion of proc macros that weirdly mix token spans (e.g. `yew::html!`)
```

>   * [#8203] fix expansion of proc macros with a float literal in a body

Reviewer: ```suggestion
  * [#8203] Fix expansion of proc macros with a float literal in a body
```

>   * [#7947] Fix exception when expanding macros in parallel and enable it by default

Reviewer: Maybe move it into `Performance Improvements` section? Also, I'd add that there was an attempt to make procedural macro expansion parallel but it was reverted because of the bug

  * [#8278] Fix procedural macro expansion with nightly toolchain

  Note, procedural macro expansion is still under development, and it's disabled by default for now.
  To turn it on, enable `org.rust.cargo.evaluate.build.scripts` and `org.rust.macros.proc` [experimental features]

> * [#8261] Fix type inference and method code completion with `sqlx` macros

Reviewer: It seems it's not only about `sqlx` lib

* [#8241] Fix handling `$crate` inside detached files

* [#8226] Don't produce [E0277] errors for non fully-known types

* [#8235] Fix false positive [E0277] `trait bound Sized is not satisfied`

* [#8235] Take into account `Sized`/`?Sized` trait bounds during name resolution and code completion

> * [#8217] Fix [`Extract function`] refactoring (<kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>M</kbd>) for argument of `println!` macro

Reviewer: ```suggestion
* [#8217] Fix [`Extract function`] refactoring (<kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>M</kbd>) for argument of some macro calls like `println!` or `vec!`
```

* [#8194] Fix `Introduce variable` intention inside match arm

> * [#8164] Fix continuous "Preparing resolve data" when memory is low

Reviewer: ```suggestion
* [#8164] Fix continuous "Preparing resolve data" process when memory is low
```

> * [#8045] Fix auto-import in `include!`-ed files

Reviewer: ```suggestion
* [#8045] Fix auto-import in `include!`-ed files. Previously, `use` items were wrongly inserted into file contained `include!` macro call
```

---
layout: post
title: "IntelliJ Rust Changelog #162"
date: 2021-12-20 13:00:00 +0300
---

## New Features

* [#8145] Take into account [const generics defaults] during const evaluation

* [#8070] Suggest private items when completing code inside `New Watch` and `Evaluate` in [Debug](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-debug-tool-window.html) tool window

* [#8192] Parse and annotate [`~const` syntax]

## Performance Improvements

* [#7947] Fix exception when expanding macros in parallel and enable it by default.

>   This is a second attempt to make procedural macro expansion parallel.

Reviewer: ```suggestion
  This is the second attempt to make procedural macro expansion parallel.
```

  The previous one was reverted because of [the bug](https://github.com/intellij-rust/intellij-rust/issues/7937).

## Fixes

* Procedural macro expansion fixes:

  * [#8271] Fix expansion of procedural macros that use `CARGO_TARGET_DIR` environment variable
    (for example, macros from [`sqlx`] crate with `offline` feature)

  * [#8233] Fix expansion of proc macros that weirdly mix token spans (e.g. `yew::html!`)

  * [#8203] Fix expansion of proc macros with a float literal in a body

  * [#8278] Fix procedural macro expansion with nightly toolchain

  Note, procedural macro expansion is still under development, and it's disabled by default for now.
  To turn it on, enable `org.rust.cargo.evaluate.build.scripts` and `org.rust.macros.proc` [experimental features]

> * [#8261] Fix type inference and method code completion with [`sqlx`] macros

Reviewer: Maybe something like
```suggestion
* [#8261] Fix type inference then the type in `impl` item is also used in associated type projection like `impl Trait1<<S as Trait2>::Item> for S {}`, In particular, it should fix type inference and code completion with [`sqlx`] macros
```
?
